### PR TITLE
model refactor tflipflop

### PR DESCRIPTION
**Pull request by Demopans**

### CODE REVIEW

**Comment by Demopans** [comment]
My IDE automatically reformat code for readability. It may not align with eslint standards

File: src/app/core/models/types/digital.ts [hunk 1]
```diff
@@ -4,33 +4,48 @@ import {DefaultWire, Wire, WireFactory}                from "./base/Wire";
...
+export type DigitalPort = Port & { kind: "DigitalPort", group: DigitalPortGroup };
+export type DigitalWire = Wire & { kind: "DigitalWire" };
 export type DigitalNode = Component & { kind: "DigitalNode" };
 
+// components
 export type ANDGate = Component & { kind: "ANDGate" };
-export type Switch  = Component & { kind: "Switch"  };
-export type LED     = Component & { kind: "LED", color: string };
+export type Switch = Component & { kind: "Switch" };
+export type LED = Component & { kind: "LED", color: string };
```
Comment (TGCrystal) [inline src/app/core/models/types/digital.ts:19]: Undo spacing changes

File: src/app/core/models/types/digital.ts [hunk 2]
```diff
@@ -4,33 +4,48 @@ import {DefaultWire, Wire, WireFactory}                from "./base/Wire";
...
+
+export type DFlipFlop = Component & { kind: "DFlipFlop" };
+export type TFlipFlop = Component & { kind: "TFlipFlop" };
+export type JKFlipFlop = Component & { kind: "JKFlipFlop" };
+export type SRFlipFlop = Component & { kind: "SRFlipFlop" };
+
+export type FlipFlop = Component &
+    (|{ kind: "DFlipFlop" }|{ kind: "TFlipFlop" }|{ kind: "JKFlipFlop" }|{ kind: "SRFlipFlop" });
```
Comment (TGCrystal) [inline src/app/core/models/types/digital.ts:27]: Wouldn't
```ts
export type FlipFlop = DFlipFlop | TFlipFlop | JKFlipFlop | SRFlipFlop;
```
work?

File: src/app/core/models/types/digital.ts [hunk 3]
```diff
@@ -4,33 +4,48 @@ import {DefaultWire, Wire, WireFactory}                from "./base/Wire";
...
 
 export type DigitalObj = DigitalPort | DigitalWire | DigitalComponent;
 
 
+/** Defines all components. */
 export const DefaultDigitalComponent: { [C in DigitalComponent as C["kind"]]: ComponentFactory<C> } = {
-    "DigitalNode": (id) => ({ ...DefaultComponent(id), kind: "DigitalNode"           }),
-    "Switch":      (id) => ({ ...DefaultComponent(id), kind: "Switch"                }),
+    "DigitalNode": (id) => ({ ...DefaultComponent(id), kind: "DigitalNode" }),
+    "Switch":      (id) => ({ ...DefaultComponent(id), kind: "Switch" }),
```
Comment (TGCrystal) [inline src/app/core/models/types/digital.ts:42]: Undo spacing changes

File: src/app/digital/models/sim/Propagators.ts [hunk 4]
```diff
@@ -19,7 +19,7 @@ type PropagatorRecord = {
  *  any state or information from the component itself.
  *
  * @param propagator The simple propagator that takes in a list of signals and outputs a list of signals.
- * @returns          The propagator function to facilitate this propagation.
+ * @returns            The propagator function to facilitate this propagation.
```
Comment (TGCrystal) [inline src/app/digital/models/sim/Propagators.ts:22]: Undo spacing change

File: src/app/digital/models/sim/Propagators.ts [hunk 5]
```diff
@@ -60,6 +60,12 @@ export const AllPropagators: PropagatorRecord = {
     "LED": Noprop,
 
     "ANDGate": InputOutputPropagator((inputs) => [inputs.reduce(AND)]),
+
+    // ToDo: Add prop here later
+    "DFlipFlop":  Noprop,
+    "TFlipFlop":  Noprop,
+    "JKFlipFlop": Noprop,
+    "SRFlipFlop": Noprop,
```
Comment (TGCrystal) [inline src/app/digital/models/sim/Propagators.ts:68]: Still needs to be done

**Review by TGCrystal — CHANGES_REQUESTED** [review]
In addition to the other things mentioned here, the functionality of the Flip Flops is still missing. Also the flip flops are missing their labels for their ports.

File: src/app/digital/models/sim/Propagators.ts (hunk 5 above)
Comment (Demopans) [reply src/app/digital/models/sim/Propagators.ts:68]: Should we have all the propagator functions in a separate file

Comment (LeonMontealegre) [reply src/app/digital/models/sim/Propagators.ts:68]: Mmmmmm, yeah, sure. You should make a `propagators` subfolders in `src/app/digital/models/sim/` and have all the FlipFlop propagators in a `FlipFlops.ts` file within that folder

**Pull request closed without merging**